pkg/endpointgc/endpointgc.go

```diff
@@ -0,0 +1,234 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package endpointgc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime/pprof"
+
+	"github.com/sirupsen/logrus"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiTypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/endpointstate"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/hive/job"
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	v2 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/promise"
+	"github.com/cilium/cilium/pkg/resiliency"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+type localEndpointCache interface {
+	LookupCEPName(namespacedName string) *endpoint.Endpoint
+}
+
+type params struct {
+	cell.In
+
+	Logger              logrus.FieldLogger
+	Lifecycle           hive.Lifecycle
+	JobRegistry         job.Registry
+	Scope               cell.Scope
+	CiliumEndpoint      resource.Resource[*types.CiliumEndpoint]
+	CiliumEndpointSlice resource.Resource[*cilium_v2a1.CiliumEndpointSlice]
+	Clientset           k8sClient.Clientset
+	RestorerPromise     promise.Promise[endpointstate.Restorer]
+	EndpointsCache      localEndpointCache
+	Cfg                 Config
+	DaemonCfg           *option.DaemonConfig
+}
+
+type gc struct {
```

Comment:
is this an actual garbage collector? I wonder whether it would help to rename it (package, struct, filename, ...) to `(Stale)EndpointCleanup` - because it only runs once at the agent startup and we should avoid confusion with the actual EndpointGC from the operator.

Reply:
Changed the package name to `endpointcleanup` to avoid confusion (also changed the filenames, the struct and so on). :+1: